### PR TITLE
[Related to JENKINS-34121] Allow disabling the protocols individually

### PR DESCRIPTION
**Pull request by stephenc**
- Also ensure that a fatal unexpected error in one protocol does not prevent a fall-back

@reviewbybees

### CODE REVIEW

File: src/main/java/hudson/remoting/Engine.java [hunk 1]
```diff
@@ -302,12 +302,24 @@ public void run() {
...
                     events.status("Trying protocol: " + protocol.getName());
                     try {
                         channel = protocol.establishChannel(jnlpSocket, channelBuilder);
                     } catch (IOException ioe) {
-                        events.status("Protocol failed to establish channel", ioe);
+                        events.status("Protocol " + protocol.getName() + " failed to establish channel", ioe);
+                    } catch (RuntimeException e) {
+                        events.status("Protocol " + protocol.getName() + " encountered a runtime error", e);
+                    } catch (Error e) {
+                        events.status("Protocol " + protocol.getName() + " could not be completed due to an error", e);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Engine.java]: Uhm. Do we really want to continue after it. Errors are commonly fatal

File: src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol.java [hunk 2]
```diff
@@ -58,6 +58,14 @@
     }
 
     /**
+     * Whether this protocol is enabled for connecting.
+     * @return {@code true} if this protocol is enabled.
+     */
+    public boolean isEnabled() {
+        return !Boolean.getBoolean(getClass().getName()+".disabled");
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol.java]: so we retrieve the property every time. Useful for debugging, performance should not suffer in this place

File: src/main/java/hudson/remoting/Engine.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/hudson/remoting/Engine.java]: IOError is one that we should continue after and I fear could be one that gets thrown

Comment (stephenc) [inline src/main/java/hudson/remoting/Engine.java]: feel free to update this PR to whatever you want, but if JnlpProtocol3 throws an Error, e.g. LinkerError, etc we should still try the others

**Comment by oleg-nenashev** [comment]
Yay, remoting build failed due to the remoting bug.

```
Tests run: 173, Failures: 0, Errors: 0, Skipped: 0

[WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: Invalid object ID 7 iota=14
Build was aborted
Putting comment on the pull request
[WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: Invalid object ID 7 iota=14
channel stopped
[WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: java.io.IOException: Backing channel is disconnected.
```

File: src/main/java/hudson/remoting/Engine.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Engine.java]: I think it's acceptable. If the error is not handleable, it will just blow up in another place

**Comment by oleg-nenashev** [comment]
🐝

**Comment by oleg-nenashev** [comment]
Created a separate issue for it: https://issues.jenkins-ci.org/browse/JENKINS-34819

**Pull request merged**